src/data/tracked-entity-instances/TEID2ApiRepository.ts

```diff
@@ -42,7 +45,16 @@ export class TEID2ApiRepository implements TEIRepository {
             return [...trackedEntityInstances, ..._.flatten(paginatedTEIs)];
         });
 
-        return result.flat();
+        return _(result)
+            .flatten()
+            .map(object => {
+                const isUpdatedAfterStartDate = new Date(object.lastUpdated).toISOString() >= startDate.format();
+                const isLastSuccessfulSync = period === "SINCE_LAST_SUCCESSFUL_SYNC";
+
+                return isUpdatedAfterStartDate || !isLastSuccessfulSync ? object : undefined;
```

Comment:
This logic is exactly the same we have in EventsD2ApiRepository.ts, right? (the block that returns a boolean) Assuming it's not accidental repetition (it does not look like it), we'd need to DRY it. Where? Maybe a function in `DataSynchronizationParams.ts`?